### DISCUSSION TITLE
Debugging and Code Failures

### DISCUSSION THREAD

**Discussion by BryanAM** [greatfrontend/greatfrontend#48] [Ideas/Feature Requests]
When a test fails, either the console or UI should indicate which test in particular failed. 
For example when Deep Equals fails at primitive, the console doesn't indicate which test failed making it hard to debug. 

Either separate the tests or improve the logging .

Full Test
```
   expect(deepEqual(0, 0)).toEqual(true);
    expect(deepEqual(-0, +0)).toEqual(false);
    expect(deepEqual('foo', 'foo')).toEqual(true);
    expect(deepEqual(true, 1)).toEqual(false);
    expect(deepEqual(true, true)).toEqual(true);
    expect(deepEqual(false, false)).toEqual(true);
    expect(deepEqual(undefined, null)).toEqual(false);
    expect(deepEqual(null, null)).toEqual(true);
    expect(deepEqual(undefined, undefined)).toEqual(true);
    expect(deepEqual(NaN, NaN)).toEqual(true);

  ``` 

logged example of an error in primitive
```
Error: expect(received).toEqual(expected) // deep equality

Expected: false
Received: true
   at @
   at @
   at Promise@[native code]
   at @
   at @
   at step@
   at @
   at Promise@[native code]
   at @
   at @
   at step@
   at @
   at Promise@[native code]
   at @
   at @
   at step@
   at @
   at Promise@[native code]
   at @
   at @
   at step@
   at @
   at Promise@[native code]
   at @
   at @
   at step@
   at @
   at Promise@[native code]
   at @
   at @
   at step@
   at @
   at Promise@[native code]
   at @
```

**Comment by yangshun** [comment]
Good point! We should add line numbers so that it's easier to tell. Fortunately it's only the type utilities question when there are many assertions within a test. Most other questions don't have so many assertions.

**Comment by akkhil7** [comment]
I think it's better if we can separate out the console.logs based on the test cases. What do you guys think?

**Comment by yangshun** [comment]
We've revamped the coding workspace and now you can see the exact line that failed <img width="751" alt="image" src="https://github.com/greatfrontend/greatfrontend/assets/1315101/3f655e75-b281-4eaf-951c-b2f2f6cdbbbd">
